### DISCUSSION TITLE
STM32 cortex-m4 使用xmake求助，紧急！

### DISCUSSION THREAD

**Discussion by geekroscom** [xmake-io/xmake#2439] [Q&A]
xmake团队的伙伴们大家好！

xmake的STM32资源比较少，在使用的过程中受阻严重，希望能够得到官方团队的支援，我们也非常希望将xmake推广到我们的开发者社群中，这事情非常有意义。

目前使用过程中参考的一些资料：
1、https://github.com/xmake-io/xmake/tree/master/tests/projects/embed/mdk/hello （示例比较简单，完整的STM32项目涉及的东西较多）
2、https://blog.csdn.net/anobodykey/article/details/80213699（参照进行了移植和配置，编译报错较多）

希望xmake团队能够基于下面相对完整的项目对STM32做下支持，以便后者参考使用。

项目源码：https://cdn.geekros.com/file/hardware.zip
MCU：STM32F427IIH6 cortex-m4
编译工具链：arm-none-eabi-gcc version 9.3.1 20200408 (release) (GNU Arm Embedded Toolchain 9-2020-q2-update)

项目目录结构：
Build：编译文件存储目录
FreeRTOS：FreeRTOS实时操作系统
Keil：Keil5项目配置文件
Source：项目源码
System：STM32F4标准库
makefile：gcc+make可用的编译配置文件，可正常生成HEX、BIN文件，可正常烧录至开发板运行
xmake.lua：xmake配置文件，已经完成add_files、remove_files、add_includedirs资源配置。

联系方式wechat：makeryang8080

希望能得到官方团队的重视，我们在STM32以及国产MCU方面有很多开发板资源，可以最大化协助xmake进行STM32的兼容适配。

再次感谢！

**Comment by geekroscom** [comment]
目前使用的构建命令：xmake f -p cross -a cortex-m4 --toolchain=armcc -c

**Comment by geekroscom** [comment]
目前遇到的报错
`shell
[ 64%]: ccache compiling.release System\usb\USB_Library\Core\src\usbd_usr.c
[ 65%]: ccache compiling.release System\usb\USB_Library\Class\cdc\src\usbd_cdc_core.c
[ 97%]: linking.release firmware.elf
error: f:/programfiles/gcc/arm-none-eabi-windows/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld.exe: f:/programfiles/gcc/arm-none-eabi-windows/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard\libc_nano.a(lib_a-sbrkr.o): in function `_sbrk_r':
sbrkr.c:(.text._sbrk_r+0xc): undefined reference to `_sbrk'
f:/programfiles/gcc/arm-none-eabi-windows/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld.exe: f:/programfiles/gcc/arm-none-eabi-windows/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard\libc_nano.a(lib_a-writer.o): in function `_write_r':
writer.c:(.text._write_r+0x10): undefined reference to `_write'
f:/programfiles/gcc/arm-none-eabi-windows/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld.exe: f:/programfiles/gcc/arm-none-eabi-windows/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard\libc_nano.a(lib_a-closer.o): in function `_close_r':
closer.c:(.text._close_r+0xc): undefined reference to `_close'
f:/programfiles/gcc/arm-none-eabi-windows/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld.exe: f:/programfiles/gcc/arm-none-eabi-windows/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard\libc_nano.a(lib_a-lseekr.o): in function `_lseek_r':
lseekr.c:(.text._lseek_r+0x10): undefined reference to `_lseek'
f:/programfiles/gcc/arm-none-eabi-windows/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld.exe: f:/programfiles/gcc/arm-none-eabi-windows/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard\libc_nano.a(lib_a-readr.o): in function `_read_r':
readr.c:(.text._read_r+0x10): undefined reference to `_read'
f:/programfiles/gcc/arm-none-eabi-windows/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld.exe: f:/programfiles/gcc/arm-none-eabi-windows/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard\libc_nano.a(lib_a-abort.o): in function `abort':
abort.c:(.text.abort+0xa): undefined reference to `_exit'
f:/programfiles/gcc/arm-none-eabi-windows/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld.exe: f:/programfiles/gcc/arm-none-eabi-windows/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard\libc_nano.a(lib_a-signalr.o): in function `_kill_r':
signalr.c:(.text._kill_r+0xe): undefined reference to `_kill'
f:/programfiles/gcc/arm-none-eabi-windows/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld.exe: f:/programfiles/gcc/arm-none-eabi-windows/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard\libc_nano.a(lib_a-signalr.o): in function `_getpid_r':
signalr.c:(.text._getpid_r+0x0): undefined reference to `_getpid'
f:/programfiles/gcc/arm-none-eabi-windows/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld.exe: f:/programfiles/gcc/arm-none-eabi-windows/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard\libc_nano.a(lib_a-fstatr.o): in function `_fstat_r':
fstatr.c:(.text._fstat_r+0xe): undefined reference to `_fstat'
f:/programfiles/gcc/arm-none-eabi-windows/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/bin/ld.exe: f:/programfiles/gcc/arm-none-eabi-windows/bin/../lib/gcc/arm-none-eabi/9.3.1/../../../../arm-none-eabi/lib/thumb/v7e-m+fp/hard\libc_nano.a(lib_a-isattyr.o): in function `_isatty_r':
isattyr.c:(.text._isatty_r+0xc): undefined reference to `_isatty'
collect2.exe: error: ld returned 1 exit status
`

**Reply by waruqi** [reply]
看着像是缺 c 库符号，可以开个 issues ，提供下  -vD 的完整编译输出，和 link commands，然后自己也可以排查下 libc 库，工具链中是否提供，可以先尝试自己 add_syslinks  加上试试

**Comment by waruqi** [comment]
> 希望xmake团队能够基于下面相对完整的项目对STM32做下支持，以便后者参考使用。

目前这边不提供 完整项目移植服务，也没时间精力搞。

> xmake的STM32资源比较少，在使用的过程中受阻严重，希望能够得到官方团队的支援，我们也非常希望将xmake推广到我们的开发者社群中，这事情非常有意义。

如果使用中遇到问题，可以正常提 issues 反馈，并提供详细错误，通常我都会回复，具体看这边的排期和空闲时间安排，毕竟还有很多 issues 和 features 需要处理，白天还得工作。

最近几天有点小忙，请耐心等待。。。

当然，我们也提供额外的技术支持，可以优先处理一些 issues 和相关咨询支持，具体见：https://github.com/xmake-io/xmake/issues/2043
（但也仅限于晚上空闲时间，不能保证完全的实时支持）

**Reply by geekroscom** [reply]
好的，不是有付费一对一支持么，没有信用卡，微信赞助了几笔。

**Reply by waruqi** [reply]
一对一支持见：https://github.com/xmake-io/xmake/issues/2043 里面说明，如果还是不清楚相关细节，可以邮件沟通 waruqi@gmail.com
